### PR TITLE
Patch `cupy` constrain

### PR DESCRIPTION
**Pull request by ajschmidt8**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This PR patches `cupy` to remove the constrain on `libcugraph` to prevent cyclical dependency issues with RAPIDS.

### CODE REVIEW

**Comment by isuruf** [comment]
How can there be cyclic dependencies with `constrains`?

**Comment by leofang** [comment]
IIUC it's cugraph->cudf->cupy->libcugraph.

File: recipe/gen_patch_json.py [hunk 1]
```diff
@@ -1187,12 +1187,16 @@ def _gen_new_index(repodata, subdir):
...
-        
+
         # auto-sklear needs to depend on the full dask.
         # https://github.com/automl/auto-sklearn/issues/1256
         if record_name == "auto-sklearn":
             _rename_dependency(fn, record, "dask-core", "dask")
 
+        # remove libcugraph constrain to prevent cyclical dependency issues with RAPIDS
+        if record_name == "cupy":
+            _replace_pin("libcugraph >=0.19.0,<1.0a0", "libcugraph", record.get("constrains", []), record, target='constrains')
```
Comment (leofang) [inline recipe/gen_patch_json.py:1198]: As discussed internally:
```suggestion
            _replace_pin("libcugraph >=0.19.0,<1.0a0", "libcugraph >=0.19.0", record.get("constrains", []), record, target='constrains')
```

**Comment by ajschmidt8** [comment]
Whoops. I accidentally removed the required review somehow :thinking:. Doesn't look like I can add it back.

**Review by rlratzel — APPROVED** [review]

**Comment by isuruf** [comment]
First, please open a PR to cupy-feedstock to fix the constraint.

File: recipe/gen_patch_json.py [hunk 2]
```diff
@@ -1187,12 +1187,16 @@ def _gen_new_index(repodata, subdir):
...
             deps = record["depends"]
             _replace_pin("google-auth >=1.25.1,<3.0dev", "google-auth >=1.25.1,<2.0dev", deps, record)
-        
+
         # auto-sklear needs to depend on the full dask.
         # https://github.com/automl/auto-sklearn/issues/1256
         if record_name == "auto-sklearn":
             _rename_dependency(fn, record, "dask-core", "dask")
 
+        # remove libcugraph constrain to prevent cyclical dependency issues with RAPIDS
```
Comment (isuruf) [inline recipe/gen_patch_json.py:1196]:
```suggestion
        # libcugraph 0.19.0 is compatible with the new calver based version 21.x
```

**Comment by ajschmidt8** [comment]
> First, please open a PR to cupy-feedstock to fix the constraint.

@isuruf, PR here: https://github.com/conda-forge/cupy-feedstock/pull/155

**Comment by jakirkham** [comment]
> > First, please open a PR to cupy-feedstock to fix the constraint.
> 
> @isuruf, PR here: [conda-forge/cupy-feedstock#155](https://github.com/conda-forge/cupy-feedstock/pull/155)

This was moved into PR ( https://github.com/conda-forge/cupy-feedstock/pull/156 ) including a recent release, which is now merged.

**Pull request merged**